include/jemalloc/internal/thread_event.h

```diff
@@ -0,0 +1,61 @@
+#ifndef JEMALLOC_INTERNAL_THREAD_EVENT_H
+#define JEMALLOC_INTERNAL_THREAD_EVENT_H
+
+#include "jemalloc/internal/tsd.h"
+
+/*
+ * Maximum threshold on thread_allocated_threshold_fast, so that there is no
+ * need to check overflow in malloc fast path. (The allocation size in malloc
+ * fast path never exceeds SC_LOOKUP_MAXCLASS.)
+ */
+#define THREAD_ALLOCATED_THRESHOLD_FAST_MAX				\
+    (UINT64_MAX - SC_LOOKUP_MAXCLASS + 1)
+
+/*
+ * Maximum threshold on wait time, so that whenever an event is triggered, the
+ * accumulated allocation bytes do not overflow uint64_t.
+ */
+#define THREAD_EVENT_WAIT_MAX (UINT64_MAX - SC_LARGE_MAXCLASS + 1)
+
+/*
+ * List of all waits, in the following format:
+ *  W(event,		(condition))
+ */
+#define ITERATE_OVER_ALL_WAITS						\
+    W(prof_sample,	(config_prof && opt_prof))
+
+void thread_event_trigger(tsd_t *tsd);
+void thread_event_update(tsd_t *tsd, uint64_t *tsd_waitp, uint64_t wait);
+
+JEMALLOC_ALWAYS_INLINE bool
+thread_event(tsd_t *tsd, size_t usize) {
+	uint64_t thread_allocated_before = tsd_thread_allocated_get(tsd);
+
+	/*
+	 * Avoid using the setter so that thread_allocated is incremented even
+	 * when tsd state is reincarnated or minimal_initialized.
+	 */
+	*tsd_thread_allocatedp_get(tsd) = thread_allocated_before + usize;
+
+	/* The subtraction is intentionally susceptible to underflow. */
+	if (likely(usize < tsd_thread_allocated_threshold_get(tsd) -
```

Comment:
nit: let's take `tsd_thread_allocated_threshold_get(tsd)` into a separate inline function, e.g. `thread_allocated_next_event`, and add some sanity checks there.

Reply:
What kind of sanity checks are you referring to?

Reply:
Assertions such as fast_threshold <= threshold, and thread.allocated_before in between of last and next threshold.